### ISSUE TITLE
500 when applying promotion on 9999 item units

### ISSUE THREAD

**Issue by tautelis** [Sylius/Sylius#10155]
**Sylius version affected**: 1.x

**Description**  
Maximum possible quantity value at sylius is 9999. So when setting this value on an item which is eligible for default Christmas promotion PromotionApplicator tries to distribute the discount for all 9999 item units which therefore generates 9999 new item unit adjustments. Persisting this many entities results in timeout and 500 server response. This place should be optimised.

**Steps to reproduce**  
* ensure there is an eligible promotion for the item in the cart
* changing item quantity to 9999 in cart edit page
* click update

**Suggestions**
There could maybe be an option if adjustments are applied on item unit or items itself. Or some sort of different logic applied by promotion applicator on high quantity items.

**Comment by kate-kate** [comment]
Hi everyone. Have found same problem. As I can see everything crushes at ```Sylius\Component\Order\Modifier\OrderItemQuantityModifier``` in increase/decreaseUntisNumber methods. First of all this place seems to be problematic for big quantity
```
    private function increaseUnitsNumber(OrderItemInterface $orderItem, int $increaseBy): void
    {
        for ($i = 0; $i < $increaseBy; ++$i) {
            $this->orderItemUnitFactory->createForItem($orderItem);
        }
    }
```

**Comment by vvasiloi** [comment]
The error is most likely caused by a timeout, i.e. reaching max execution time, or reaching the memory limit.
I was able to add update the quantity of a cart item from 1 to 1000, thus adding 9999 more order item units.
Here's a view from the profiler:

![performance](https://user-images.githubusercontent.com/7114562/62369876-b242db00-b539-11e9-9f2b-464ceab772db.png)

![database](https://user-images.githubusercontent.com/7114562/62369882-b5d66200-b539-11e9-858c-1ec9ba13c771.png)

**Comment by itinance** [comment]
Do we really need this OrderItemUnit? In my case there is a digital product that could cause orders of 1.000.000.000 pieces which will not be possible with such a loop.

**Comment by vvasiloi** [comment]
@itinance promotions, tax calculation, shipments, etc.
It represents a single unit that is bought.
If you don't need it in your project, you could probably only create one OrderItemUnit per OrderItem, just like Sylius by default creates only one shipment.
Responsible for item generation is `OrderItemQuantityModifier`.

**Comment by itinance** [comment]
Thx for clarification @vvasiloi.

**Comment by itinance** [comment]
It turns out, that we have a serious issue here. The total amount of items in an Order is reflected by the number of `OrderItemUnits`.

Overriding the internal `OrderItemQuantityModifier` like this:

```
    sylius.order_item_quantity_modifier:
        class: AppBundle\Component\Core\Cart\Modifier\MyOrderItemQuantityModifier
        public: true
        arguments:
            - "@sylius.factory.order_item_unit"
```

with a custom class, where we don't create 1 million of OrderItemUnit-Records when a Order was placed with 1 Million Items of digital goods, then either an Exception is thrown for Tax-calculation stuff or - if we create at least 1 OrderUnitItem, the final amount is 100 instead of a million.

As we don't sell Milk or Clothes, but digital goods in a possible amount of millions of items, this kind of Order management is not suitable for us. Any ideas how to make this work with Sylius having not millions of records for a single order of 1 million items?

**Comment by vvasiloi** [comment]
> One approach could be to not create a single OrderUnitItem for each item

Isn't that what I said?

> If you don't need it in your project, you could probably only create one OrderItemUnit per OrderItem, just like Sylius by default creates only one shipment.

> but add a "amount"-property into the UnitItem for reflecting a chunk of multiple items instead only one.

There's already a `quantity` field on the OrderItem, why do you need one more on the OrderItemUnit?

**Comment by itinance** [comment]
Hey @vvasiloi 

> There's already a `quantity` field on the OrderItem, why do you need one more on the OrderItemUnit?

As I see in the source code, `addUnit()` is required in order to increase the quantity of an `OrderItem`.  `OrderItem` itself doesn't have a `setQuantity`-Method or any other thing to set the value manually. If it would have one, we could skip the OrderItemUnits.

Do you have any other suggestion instead of updating the `sylius_order_item`-table manually per SQL?


`OrderItem.php`:

```
    public function addUnit(OrderItemUnitInterface $unit): void
    {
        if ($this !== $unit->getOrderItem()) {
            throw new \LogicException('This order item unit is assigned to a different order item.');
        }

        if (!$this->hasUnit($unit)) {
            $this->units->add($unit);

            ++$this->quantity;
            $this->unitsTotal += $unit->getTotal();
            $this->recalculateTotal();
        }
    }
```

**Comment by vvasiloi** [comment]
@itinance override the method, remove the statement that increase the quantity and add setter/getter for quantity?

**Comment by itinance** [comment]
Apparently this is not enough. The `total` in the balance sheet / order sheet will be 0 at all.

What I did:

**1. Override `order_item_quantity_modifier`:**

```yml
    sylius.order_item_quantity_modifier:
        class: AppBundle\Component\Core\Cart\Modifier\MyOrderItemQuantityModifier
        public: true
        arguments:
            - "@sylius.factory.order_item_unit"

```

with this:

```php
class MyOrderItemQuantityModifier implements OrderItemQuantityModifierInterface
{
	/** @var OrderItemUnitFactoryInterface */
	private $orderItemUnitFactory;

	public function __construct(OrderItemUnitFactoryInterface $orderItemUnitFactory)
	{
		$this->orderItemUnitFactory = $orderItemUnitFactory;
	}

	/**
	 * {@inheritdoc}
	 */
	public function modify(OrderItemInterface $orderItem, int $targetQuantity): void
	{
		if($orderItem instanceof OrderItem) {
			$orderItem->setQuantity($targetQuantity);
		}
	}
}
```

**2. Override OrderItem:**

```yml
sylius_order:
    resources:
        order_item:
            classes:
                model: AppBundle\Entity\OrderItem
```

Class:

```php
<?php

namespace AppBundle\Entity;

use Sylius\Component\Core\Model\OrderItem as BaseOrderItem;
use Sylius\Component\Order\Model\OrderItemUnitInterface;

class OrderItem extends BaseOrderItem
{
	/**
	 * @param int $quantity
	 * @return BaseOrderItem
	 */
	public function setQuantity(int $quantity) : void {
		$this->quantity += $quantity;
		$this->unitsTotal += $quantity;

		$this->recalculateTotal();
	}
}

```

Doctrine `OrderItem.orm.yml`:

```yml
AppBundle\Entity\OrderItem:
  type: entity
  table: sylius_order_item
```

After submitting a new order of lets say 5000 items, the following is happening at the end in `sylius_order_item`:

> quantity: 5000
> units_total: 0
> total: 0
> 

When I `dump` the order-item within the `OrderItemQuantityModifier`, everything seems to be okay having valid values on all the properties.

But after the order was processed, `total` and `units_total` tends to zero, which is shown in the order-overview also.

I miss something here.

**Comment by vvasiloi** [comment]
@itinance 
1. You need at least one order item unit, as mentioned before
2. `$this->unitsTotal ` should be `$quantity * $unit->getTotal()` or something like that, it's not the amount of units (that's quantity), but the units total price.
You might need to do more changes in the code, including the `addUnit` method, to remove the old logic.

**Comment by tautelis** [comment]
@Zales0123 any progress on this Sylius issue? anything in the roadmap?

**Comment by antiseptikk** [comment]
Hi everyone.
Have you found a solution?

It's a big problem for me, in my case I have to add a lot of products to the cart and if the total is more than 1k or 2k it's really slow and 500 are thrown out.

**Comment by einorler** [comment]
Hey, guys, this is really a huge problem for non-standard shops that deal with products like parts, paper, or anything that has to do with large amounts of products being sold @Zales0123 @pjedrzejewski , is there anything that can be done about this?

**Comment by lchrusciel** [comment]
Hey folks, unfortunately, we probably cannot deal with it before 1.8. We will try to prioritize it after nearest release

**Comment by ddanninger** [comment]
> Hey folks, unfortunately, we probably cannot deal with it before 1.8. We will try to prioritize it after nearest release

Did this by any chance get fixed in the 1.8 release?

**Comment by vvasiloi** [comment]
@ddanninger no.

**Comment by einorler** [comment]
sooo... Having in mind what Ichrusciel said on July the 20th, hows the prioritization going? :)

**Comment by ddanninger** [comment]
How can we help you to find a solution for this?

**Comment by matjazbo** [comment]
I think solving tax/promotion/vat and other calculations by creating a new row in database table for each unit is a bad design decision at least it does not scale to bigger quantities. Also it prevents having decimal values for quantity, which basically prevents any other units of meassure than pieces.

This seriously limits use cases for Sylius.

**Comment by vvasiloi** [comment]
@matjazbo you can convert decimal numbers to integers to a certain point, like it's currently done for prices, or add custom data points to store float values, so that's not an issue.

**Comment by matjazbo** [comment]
@vvasiloi you're right, but you're still limited to this 9999 max quantity hard cap, meaning that when for example 2 decimal points are introduced, the maximum quantity for such an item is around 99.99

**Comment by vvasiloi** [comment]
@matjazbo this limit can be changed by overriding the `sylius.order_item_quantity_modifier.limiting` service and changing the 2nd parameter, which by default is 9999. From version 1.9 this can be done through the newly introduced `sylius.order_item_quantity_modifier.limit` parameter.

**Comment by matjazbo** [comment]
@vvasiloi I'm not talking about the parameter limit, but the limit of Sylius handling big quantities, one example #10741.
Even if you remove this parameter limit, you're still limited because Sylius creates new database object for every quantity unit. This is the real problem.

**Comment by vvasiloi** [comment]
@matjazbo Yes, of course. This issue is difficult to solve and will most likely involve BC breaks.

**Comment by fnowacki** [comment]
@lchrusciel 
> Hey folks, unfortunately, we probably cannot deal with it before 1.8. We will try to prioritize it after nearest release

Any updates? 🤞🏻

**Comment by jakubtobiasz** [comment]
Hello @tautelis 👋🏼,
we've checked this on our demo (`https://demo.sylius.com`, Sylius `1.12.9`) and applying christmas promo code on a cart with 9999 items worked fine. If you're still struggling with this issue, some performance optimizations on your site might be required.

We'll cover such topics like optimization & performance in our new docs.

**Comment by fnowacki** [comment]
@jakubtobiasz for higher quantities the problem still occurs - for example 25000. The way you create units for each single quantity might be a performance issue

**Comment by jakubtobiasz** [comment]
Ok, so I'm reopening the issue, and hopefully, we'll find some space to check it out :). Thank you very much for letting me know the issue continues to occur 👊🏼!

**Comment by TheMilek** [comment]
Hi!

I've just taken another look at this issue. Handling 9,999 units still works correctly on our demo: https://v2.demo.sylius.com/.

While it may still occur with significantly larger quantities - like said @fnowacki, we've also documented a performance optimization approach for use cases requiring much larger quantities, based on simplifying the order structure from three layers to two:

https://docs.sylius.com/the-performance-guide/performance-optimization-concept

Given the documented optimization approach for larger quantities, I think we can close this issue. Feel free to reopen it with additional details if you're still experiencing performance issues that aren't addressed by this approach 🚀
